lib/fencing/st_client.c

```diff
@@ -2677,3 +2672,152 @@ i_hate_pils(int rc)
+    if (strstr(line, "invalid") || strstr(line, "variable")) {
+        crm_debug("Skipping: %s", line);
+        return;
+    }
+
+    crm_trace("Processing %d bytes: [%s]", max, line);
+    /* Skip initial whitespace */
+    for (lpc = 0; lpc <= max && isspace(line[lpc]); lpc++) {
+        last = lpc + 1;
+    }
+
+    /* Now the actual content */
+    for (lpc = 0; lpc <= max; lpc++) {
+        gboolean a_space = isspace(line[lpc]);
+
+        if (a_space && lpc < max && isspace(line[lpc + 1])) {
+            /* fast-forward to the end of the spaces */
+
+        } else if (a_space || line[lpc] == ',' || line[lpc] == ';' || line[lpc] == 0) {
+            int rc = 1;
+            char *entry = NULL;
+
+            if (lpc != last) {
+                entry = calloc(1, 1 + lpc - last);
+                rc = sscanf(line + last, "%[a-zA-Z0-9_-.]", entry);
+            }
+
+            if (entry == NULL) {
+                /* Skip */
+            } else if (rc != 1) {
+                crm_warn("Could not parse (%d %d): %s", last, lpc, line + last);
+            } else if (safe_str_neq(entry, "on") && safe_str_neq(entry, "off")) {
+                crm_trace("Adding '%s'", entry);
+                *output = g_list_append(*output, entry);
+                entry = NULL;
+            }
+
+            free(entry);
+            last = lpc + 1;
+        }
+    }
+}
+
+GListPtr
+parse_host_list(const char *hosts)
+{
+    int lpc = 0;
+    int max = 0;
+    int last = 0;
+    GListPtr output = NULL;
+
+    if (hosts == NULL) {
+        return output;
+    }
+
+    max = strlen(hosts);
+    for (lpc = 0; lpc <= max; lpc++) {
+        if (hosts[lpc] == '\n' || hosts[lpc] == 0) {
+            char *line = NULL;
+            int len = lpc - last;
+
+            if(len > 1) {
+                line = malloc(1 + len);
+            }
+
+            if(line) {
+                snprintf(line, 1 + len, "%s", hosts + last);
+                line[len] = 0; /* Because it might be '\n' */
+                parse_host_line(line, len, &output);
+                free(line);
+            }
+
+            last = lpc + 1;
+        }
+    }
+
+    crm_trace("Parsed %d entries from '%s'", g_list_length(output), hosts);
+    return output;
+}
+
+gboolean
+string_in_list(GListPtr list, const char *item)
+{
+    int lpc = 0;
+    int max = g_list_length(list);
+
+    for (lpc = 0; lpc < max; lpc++) {
+        const char *value = g_list_nth_data(list, lpc);
+
+        if (safe_str_eq(item, value)) {
+            return TRUE;
+        } else {
+            crm_trace("%d: '%s' != '%s'", lpc, item, value);
+        }
+    }
+    return FALSE;
+}
+
+gboolean
+watchdog_fencing_enabled_for_node(const char *node)
+{
+    gboolean rv = FALSE;
+    stonith_t *stonith_api = stonith_api_new();
+    char *list = NULL;
+
+    if(stonith_api) {
+        int rc = stonith_api->cmds->connect(stonith_api, "stonith-api", NULL);
```

Comment:
It would be more efficient to allow the caller (i.e. crmd) to pass in an active connection, and do this only if it is NULL.